File: typescript/src/main/vsdk/toolkit/scene/VoxelData.ts
import { Geometry } from "../skin/Geometry";
import { Patch } from "../skin/Patch";
import { VoxelDataFlags } from "./VoxelDataFlags";
import type { VoxelGrid } from "./VoxelGrid";

export class VoxelData {
  public patch: Patch | null;
  public geometry: Geometry | null;
  public voxelGrid: VoxelGrid | null;
  public flags: number;

  public constructor(data: VoxelGrid, flags: number);
  public constructor(data: Patch, flags: number);
  public constructor(data: Geometry, flags: number);
  public constructor(data: VoxelGrid | Patch | Geometry, flags: number) {
    this.geometry = null;
    this.patch = null;
    this.voxelGrid = null;
    this.flags = flags;

    if ("gridIntersect" in (data as object)) {
      this.voxelGrid = data as VoxelGrid;
    }
    else if ("numberOfVertices" in (data as object)) {
      this.patch = data as Patch;
    }
    else {
      this.geometry = data as Geometry;
    }
  }

  public updateRayId(id: number): void {
    this.flags = (this.flags & ~VoxelDataFlags.VOXEL_DATA_RAY_COUNT_MASK) | (id & VoxelDataFlags.VOXEL_DATA_RAY_COUNT_MASK);
  }

  public lastRayId(): number {
    return this.flags & VoxelDataFlags.VOXEL_DATA_RAY_COUNT_MASK;
  }

  public isPatch(): boolean {
    return (this.flags & VoxelDataFlags.VOXEL_DATA_PATCH_MASK) !== 0;
  }

  public isGeom(): boolean {
    return (this.flags & VoxelDataFlags.VOXEL_DATA_GEOMETRY_MASK) !== 0;
  }

  public isGrid(): boolean {
    return (this.flags & VoxelDataFlags.VOXEL_DATA_GRID_MASK) !== 0;
  }
}

File: java/src/main/java/vsdk/toolkit/scene/VoxelData.java
package vsdk.toolkit.scene;

import vsdk.toolkit.skin.Geometry;
import vsdk.toolkit.environment.geometry.elements.Patch;

class VoxelData {
    Patch patch;
    Geometry geometry;
    VoxelGrid voxelGrid;

    int flags; // Patch or geometry? last ray id, ...

    VoxelData(VoxelGrid data, int flags) {
        this.geometry = null;
        this.patch = null;
        this.voxelGrid = data;
        this.flags = flags;
    }

    VoxelData(Patch data, int flags) {
        this.geometry = null;
        this.patch = data;
        this.voxelGrid = null;
        this.flags = flags;
    }

    VoxelData(Geometry data, int flags) {
        this.geometry = data;
        this.patch = null;
        this.voxelGrid = null;
        this.flags = flags;
    }

    void updateRayId(final int id) {
        flags = (flags & ~VoxelDataFlags.VOXEL_DATA_RAY_COUNT_MASK) | (id & VoxelDataFlags.VOXEL_DATA_RAY_COUNT_MASK);
    }

    int lastRayId() {
        return flags & VoxelDataFlags.VOXEL_DATA_RAY_COUNT_MASK;
    }

    boolean isPatch() {
        return (flags & VoxelDataFlags.VOXEL_DATA_PATCH_MASK) != 0;
    }

    boolean isGeom() {
        return (flags & VoxelDataFlags.VOXEL_DATA_GEOMETRY_MASK) != 0;
    }

    boolean isGrid() {
        return (flags & VoxelDataFlags.VOXEL_DATA_GRID_MASK) != 0;
    }
}
